File: easy/589_n_ary_tree_preorder_traversal/main.rb
#
# @lc app=leetcode id=589 lang=ruby
#
# [589] N-ary Tree Preorder Traversal
#

# @lc code=start
# Definition for a Node.
class Node
  attr_accessor :val, :children

  def initialize(val)
    @val = val
    @children = []
  end
end

# @param {Node} root
# @return {List[int]}
def preorder(root)
  dfs(root, [])
end

def dfs(node, ary)
  return ary if node.nil?

  ary << node.val
  node.children.each { dfs(_1, ary) }
  ary
end
# @lc code=end

# root = [1,null,3,2,4,null,5,6]
root = Node.new(1).tap do |n|
  n.children << Node.new(3).tap do |n3|
    n3.children << Node.new(5)
    n3.children << Node.new(6)
  end
  n.children << Node.new(2)
  n.children << Node.new(4)
end
p preorder(root)

File: easy/589_n_ary_tree_preorder_traversal/main.go
package main

import "fmt"

type Node struct {
	Val      int
	Children []*Node
}

func postorder(root *Node) []int {
	result := []int{}
	dfs(root, &result)
	return result
}

func dfs(node *Node, result *[]int) {
	if node == nil {
		return
	}

	for _, n := range node.Children {
		dfs(n, result)
	}
	*result = append(*result, node.Val)
}

func main() {
	// root = [1,null,3,2,4,null,5,6]
	root := &Node{
		Val: 1,
		Children: []*Node{
			{
				Val: 3,
				Children: []*Node{
					{
						Val: 5,
					},
					{
						Val: 6,
					},
				},
			},
			{
				Val: 2,
			},
			{
				Val: 4,
			},
		},
	}

	fmt.Println(postorder(root))
}
